src/keys/mod.rs

```diff
@@ -299,6 +299,69 @@ pub trait ToDescriptorKey<Ctx: ScriptContext>: Sized {
     fn to_descriptor_key(self) -> Result<DescriptorKey<Ctx>, KeyError>;
 }
 
+/// Enum for extended keys that can be either `xprv` or `xpub`
+///
+/// An instance of [`ExtendedKey`] can be constructed from an [`ExtendedPrivKey`](bip32::ExtendedPrivKey)
+/// or an [`ExtendedPubKey`](bip32::ExtendedPubKey) by using the `From` trait.
+///
+/// Defaults to the [`Legacy`](miniscript::Legacy) context.
+pub enum ExtendedKey<Ctx: ScriptContext = miniscript::Legacy> {
+    /// A private extended key, aka an `xprv`
+    Private((bip32::ExtendedPrivKey, PhantomData<Ctx>)),
+    /// A public extended key, aka an `xpub`
+    Public((bip32::ExtendedPubKey, PhantomData<Ctx>)),
+}
+
+impl<Ctx: ScriptContext> ExtendedKey<Ctx> {
+    /// Return whether or not the key contains the private data
+    pub fn has_secret(&self) -> bool {
+        match self {
+            ExtendedKey::Private(_) => true,
+            ExtendedKey::Public(_) => false,
+        }
```

Comment:
Might be slightly nicer to use `matches!`, depending on your point of view.

Reply:
> Networks (mainnet etc) are fast becoming one of my least liked parts of cryptocurrency software, code that deals with them is almost always contrived. It would be nice to improve this valid network stuff at some stage if we can ...

Do you mean you'd like to avoid validating against specific blockchains such as Bitcoin or just make it easier to expend the possibly instances of the Bitcoin blockchain such as to support new custom signet networks? 

I do agree it would be nice to be able to easily use a custom genesis block for testing against custom signet networks which core and some other projects are beginning to use.

Reply:
> Do you mean you'd like to avoid validating against specific blockchains such as Bitcoin or just make it easier to expand the possibly instances of the Bitcoin blockchain such as to support new custom signet networks?

I'm not well versed enough in the bdk codebase to be making any claims at the moment :) I was merely commenting that including the network in code seems to always lead to contrived code. My personal opinion is that software shouldn't care or know which network it is attached too. If the point of testnets is to test the code then there should be no difference in the code irrespective of the network, if the code is different depending on the network then we are not testing the same code that would run when connected to mainnet. I'm only 18 months deep in cryptocurrency software so don't put too much weight in my opinions but this came up last week in my day job too (and it happened in my last job too) so when I saw it again so soon in bdk I felt like commenting :)

Reply:
I think the main point against reusing the same encoding for everything is to make sure users don't accidentally mix up addresses/keys/etc.

One thing I'd like to have is an extended key format that has a *set* of valid networks, or at least doesn't include one at all: most of the "hacks" we have to do in BDK come from the fact that the "base type" we reduce everything to is `ExtendedPrivKey` (or `ExtendedPubKey`) and they both contain a single network.

So when you take a different key type, like a mnemonic, and try to convert it to the "base type" you are forced to pick one single network, which means that you have to use a random one as a placeholder but also remember that the original key type didn't encode any network information, so while the key inside says that it's i.e. a `Testnet` key, it should also be considered valid on `Mainnet`.

Also it doesn't help that different key formats encode different information inside: some don't encode anything at all, just the key, (BIP39), some encode the network but not the type of script (`xprv`/`tprv`), others include whether the script is legacy or segwit as well (Electrum mnemonics).